### PR TITLE
Reduce memory overhead when creating Micrometer timers

### PR DESCRIPTION
**Pull request by pderop**
The purpose of this PR is to reduce memory overhead when creating Micrometer timers instances.

### CODE REVIEW

**Comment by violetagg** [comment]
Related to #1952

File: reactor-netty-http/src/main/java/reactor/netty/http/client/MicrometerHttpClientMetricsRecorder.java [hunk 1]
```diff
@@ -80,14 +81,18 @@ public void recordDataSentTime(SocketAddress remoteAddress, String uri, String m
...
-				                   .tags(REMOTE_ADDRESS, address, URI, uri, METHOD, method, STATUS, status)
-				                   .register(REGISTRY)));
+		Timer responseTime = getResponseTimeTimer(name() + RESPONSE_TIME, address, uri, method, status);
 		if (responseTime != null) {
 			responseTime.record(time);
 		}
 	}
+
+	@Nullable
+	Timer getResponseTimeTimer(String name, String address, String uri, String method, String status) {
```
Comment (violetagg) [inline reactor-netty-http/src/main/java/reactor/netty/http/client/MicrometerHttpClientMetricsRecorder.java:91]: What do you think to make this `final`?

File: reactor-netty-core/src/main/java/reactor/netty/transport/MicrometerAddressResolverGroupMetrics.java [hunk 2]
```diff
@@ -73,15 +74,22 @@
...
 		final String protocol;
 		final String remoteAddress;
+		final MicrometerChannelMetricsRecorder recorder;
 
 		// status is not known beforehand
 		String status;
 
-		FutureHandlerContext(String protocol, String remoteAddress) {
+		FutureHandlerContext(MicrometerChannelMetricsRecorder recorder, String protocol, String remoteAddress) {
+			this.recorder = recorder;
```
[the remark this answers is not in the document]
Comment (pderop) [reply reactor-netty-core/src/main/java/reactor/netty/transport/MicrometerAddressResolverGroupMetrics.java:82]: I applied your feedbacks, thanks.

**Comment by violetagg** [comment]
Rebased in order to pick up #2070

**Comment by pderop** [comment]
there is a compile issue, the main branch does not seem to compile ?

> Task :reactor-netty-core:compileJava
/Users/pderop/Work/projects/reactor-netty-pderop/reactor-netty-core/src/main/java/reactor/netty/transport/ByteBufAllocatorMetrics.java:18: error: package io.micrometer.core.instrument does not exist
import io.micrometer.core.instrument.Gauge;

**Comment by pderop** [comment]
> Task :reactor-netty-core:compileJava
/Users/pderop/Work/projects/reactor-netty-pderop/reactor-netty-core/src/main/java/reactor/netty/transport/ByteBufAllocatorMetrics.java:18: error: package io.micrometer.core.instrument does not exist
import io.micrometer.core.instrument.Gauge;
```

**Comment by pderop** [comment]
it seems that the main branch does not build ? I have some compilation errors, like:

```
> Task :reactor-netty-core:compileJava
/Users/pderop/Work/projects/reactor-netty-pderop/reactor-netty-core/src/main/java/reactor/netty/transport/ByteBufAllocatorMetrics.java:18: error: package io.micrometer.core.instrument does not exist
import io.micrometer.core.instrument.Gauge;
```

**Comment by pderop** [comment]
hold on, maybe I have an old snapshot in my ~/.m2/reposiry ...

**Comment by violetagg** [comment]
the build on the CI is passing

**Comment by pderop** [comment]
my bad, I previously compiled an older version of micrometer libraries, I'm finishing the rebase ...

**Comment by pderop** [comment]
@violetagg , can you review again , I think it's now ok ? thanks

File: reactor-netty-core/src/main/java/reactor/netty/Metrics.java [hunk 3]
```diff
@@ -30,7 +31,7 @@
 public class Metrics {
 	public static final MeterRegistry REGISTRY = io.micrometer.core.instrument.Metrics.globalRegistry;
 	static {
-		REGISTRY.withTimerObservationHandler();
+		REGISTRY.observationConfig().observationHandler(new ReactorNettyTimerObservationHandler(REGISTRY));
```
Comment (violetagg) [inline reactor-netty-core/src/main/java/reactor/netty/Metrics.java:34]: Discussed the change with @marcingrzejszczak, we should the change below


```suggestion
		REGISTRY.observationConfig()
				.observationHandler(new ObservationHandler.FirstMatchingCompositeObservationHandler(
						new ReactorNettyTimerObservationHandler(REGISTRY), new TimerObservationHandler(REGISTRY)));
```

File: reactor-netty-core/src/main/java/reactor/netty/Metrics.java [hunk 4]
```diff
@@ -16,6 +16,7 @@
 package reactor.netty;
 
 import io.micrometer.core.instrument.MeterRegistry;
+import reactor.netty.observability.ReactorNettyTimerObservationHandler;
```
Comment (violetagg) [inline reactor-netty-core/src/main/java/reactor/netty/Metrics.java:19]:
```suggestion
import io.micrometer.core.instrument.observation.ObservationHandler;
import io.micrometer.core.instrument.observation.TimerObservationHandler;
import reactor.netty.observability.ReactorNettyTimerObservationHandler;
```

**Review by violetagg — CHANGES_REQUESTED** [review]

**Review by violetagg — APPROVED** [review]

**Comment by violetagg** [comment]
@simonbasle PTAL

**Review by simonbasle — APPROVED** [review]

**Pull request merged**